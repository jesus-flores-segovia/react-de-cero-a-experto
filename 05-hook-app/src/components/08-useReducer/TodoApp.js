import React, { useEffect, useReducer } from 'react';
import {todoReducer} from './todoReducer';

import './styles.css';
import { TodoList } from './components/TodoList';
import { TodoAdd } from './components/TodoAdd';

const init = () => {
    return JSON.parse(localStorage.getItem("todos")) || [];
}

export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem("todos", JSON.stringify(todos));
    }, [todos]);

    const handleAdd = (description) => {

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        dispatch({
            type: "add",
            payload: newTodo
        });
    };

    const handleDelete = (todoId) => {

        dispatch({
            type: "delete",
            payload: todoId
        });
    };

    const handleToggle = (todoId) => {
        
        dispatch({
            type: "toggle",
            payload: todoId
        });
    };

    return (
        <div>
            <h1>TodoApp ( {todos.length} )</h1>
            <hr/>

            <div className="row">
                <div className="col-7">
                    <TodoList todos={todos} handleDelete={handleDelete} handleToggle={handleToggle}/>
                </div>

                <div className="col-5">
                    <TodoAdd handleAdd={handleAdd}/>
                </div>
            </div>
        </div>
    )
}
