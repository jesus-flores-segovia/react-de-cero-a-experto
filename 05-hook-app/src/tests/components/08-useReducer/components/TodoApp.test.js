import { act } from "@testing-library/react";
import { mount, shallow } from "enzyme";
import { TodoApp } from "../../../../components/08-useReducer/TodoApp";
import { todos } from "../fixtures/todos";

describe("Tests inside file 'TodoApp.test.js'", () => {
    
    const wrapper = shallow(<TodoApp/>)

    Storage.prototype.setItem = jest.fn(() => {});

    test("'TodoApp' component must show correctly", () => {
        expect(wrapper).toMatchSnapshot();
    });

    test("'TodoApp' component must add two new todos", () => {
        const wrapper = mount(<TodoApp/>);

        act(() => {
            wrapper.find("TodoAdd").prop("handleAdd")(todos[0].desc);
            wrapper.find("TodoAdd").prop("handleAdd")(todos[1].desc);
        });

        expect(wrapper.find("h1").text()).toBe("TodoApp ( 2 )");
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    });

})
