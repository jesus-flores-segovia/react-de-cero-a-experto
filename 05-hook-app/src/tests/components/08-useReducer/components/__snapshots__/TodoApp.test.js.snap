// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests inside file 'TodoApp.test.js' 'TodoApp' component must show correctly 1`] = `
<div>
  <h1>
    TodoApp ( 
    0
     )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAdd={[Function]}
      />
    </div>
  </div>
</div>
`;
